File: samples/profile/competitors/java/codec/JsonDom.java
// Java codec competitor — parse the nativejson corpus to a DOM (Jackson
// JsonNode tree) and emit one schema-conformant CSV row per (file, library),
// matching competitors/columns.txt. Library: jackson-databind, the de-facto
// standard JVM JSON library (the JDK has no stdlib JSON). Cross-check: a known
// element count per file (twitter.statuses=100 / citm.events=184 /
// canada.features=1). Jackson jars are vendored under competitors/vendor/jackson
// and placed on the classpath by codec.sh.
//
// Notes: timings include JVM JIT warmup (honoring PROFILE_WARMUP/TRIALS like the
// other runners; readTree over these corpora is large enough to OSR-compile in a
// trial). alloc_bytes is per-thread bytes allocated for one parse, via the
// HotSpot com.sun.management.ThreadMXBean — the Java-native analog of Go's
// TotalAlloc delta / Rust's global allocator / C++ operator new override.
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.sun.management.ThreadMXBean;
import java.lang.management.ManagementFactory;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Arrays;

public class JsonDom {
    static final ThreadMXBean TMX = (ThreadMXBean) ManagementFactory.getThreadMXBean();

    // (dataset_name, filename, expected count)
    static final String[][] CORPUS = {
        {"twitter", "twitter.json", "100"},
        {"citm_catalog", "citm_catalog.json", "184"},
        {"canada", "canada.json", "1"},
    };

    static String env(String k, String d) { String v = System.getenv(k); return (v == null || v.isEmpty()) ? d : v; }

    static long peakRssKb() {
        try {
            for (String line : Files.readAllLines(Path.of("/proc/self/status")))
                if (line.startsWith("VmHWM:")) return Long.parseLong(line.replaceAll("[^0-9]", ""));
        } catch (Exception e) { }
        return -1;
    }

    static long allocOf(Runnable r) {
        long id = Thread.currentThread().threadId();
        long b = TMX.getThreadAllocatedBytes(id);
        r.run();
        return TMX.getThreadAllocatedBytes(id) - b;
    }

    static int count(JsonNode node, String dataset) {
        if (node == null) return -1;
        JsonNode child = switch (dataset) {
            case "twitter" -> node.get("statuses");
            case "citm_catalog" -> node.get("events");
            case "canada" -> node.get("features");
            default -> null;
        };
        return child == null ? -1 : child.size();
    }

    static void emit(String runId, String ts, String dataset, long bytes, String lib, String ver,
                     int warmup, int trials, long[] samples, boolean ok, long alloc) {
        Arrays.sort(samples);
        int n = samples.length;
        long sum = 0; for (long x : samples) sum += x;
        int idx = Math.min(n - 1, n * 95 / 100);
        long mn = samples[0], med = samples[n / 2], mean = sum / n, p95 = samples[idx];
        double mbps = med > 0 ? (double) bytes / (double) med * 1e9 / 1048576.0 : 0.0;
        String status = ok ? "ok" : "invalid";
        System.out.printf(
            "1,%s,%s,json-dom,codec,%s,%d,,%d,java,%s,%s,%s,-server,%d,%d,%d,%d,%d,%d,%.1f,MB/s,%d,-1,%d,-1,-1,%s,%b,,%n",
            runId, ts, dataset, bytes, bytes, env("PROFILE_LANG_VERSION", ""), lib, ver,
            warmup, trials, mn, med, mean, p95, mbps, peakRssKb(), alloc, status, ok);
    }

    public static void main(String[] args) throws Exception {
        String dataDir = System.getenv("PROFILE_DATA_DIR");
        if (dataDir == null || dataDir.isEmpty()) {
            System.err.println("java/codec: PROFILE_DATA_DIR unset — skipping");
            return;
        }
        String runId = env("PROFILE_RUN_ID", "local");
        String ts = env("PROFILE_RUN_TS", "");
        int warmup = Integer.parseInt(env("PROFILE_WARMUP", "3"));
        int trials = Integer.parseInt(env("PROFILE_TRIALS", "10"));
        String ver = env("PROFILE_JACKSON_VERSION", "2.22.0");

        ObjectMapper mapper = new ObjectMapper(); // reused across trials (its intended use)

        for (String[] c : CORPUS) {
            String dataset = c[0], fname = c[1];
            int expect = Integer.parseInt(c[2]);
            Path path = Path.of(dataDir, fname);
            if (!Files.exists(path)) continue; // dataset absent → skip
            byte[] raw = Files.readAllBytes(path);
            long nbytes = raw.length;

            boolean check;
            try {
                check = count(mapper.readTree(raw), dataset) == expect;
            } catch (Exception e) {
                check = false;
            }
            long alloc = allocOf(() -> { try { mapper.readTree(raw); } catch (Exception e) { } });
            for (int i = 0; i < warmup; i++) mapper.readTree(raw);
            long[] samples = new long[trials];
            for (int i = 0; i < trials; i++) {
                long t0 = System.nanoTime();
                JsonNode node = mapper.readTree(raw);
                samples[i] = System.nanoTime() - t0;
                if (node == null) check = false;
            }
            emit(runId, ts, dataset, nbytes, "jackson-databind", ver, warmup, trials,
                 samples, check, alloc);
        }
    }
}

File: samples/profile/competitors/python/codec/json_dom.py
#!/usr/bin/env python3
# Python codec competitors — parse the nativejson corpus to a Python object DOM
# and emit one schema-conformant CSV row per (file, library), matching
# competitors/columns.txt. Libraries: json (stdlib), orjson, msgspec, ujson.
# A library that isn't importable is recorded as a skip row (never a crash).
# Cross-check: a known element count per file.
import importlib
import os
import resource
import sys
import time
import tracemalloc

FILES = [
    ("twitter", "twitter.json", 100),
    ("citm_catalog", "citm_catalog.json", 184),
    ("canada", "canada.json", 1),
]


def env(k, d):
    v = os.environ.get(k)
    return v if v else d


def peak_rss_kb():
    # ru_maxrss is KB on Linux.
    return resource.getrusage(resource.RUSAGE_SELF).ru_maxrss


_ALLOC = -1


def alloc_bytes(run):
    """Peak Python-heap bytes allocated by one execution (tracemalloc). Note: C-
    extension allocations outside CPython (e.g. numpy buffers) are not traced."""
    tracemalloc.start()
    tracemalloc.reset_peak()
    run()
    _peak = tracemalloc.get_traced_memory()[1]
    tracemalloc.stop()
    return _peak


def count(v, dataset):
    try:
        if dataset == "twitter":
            return len(v["statuses"])
        if dataset == "citm_catalog":
            return len(v["events"])
        if dataset == "canada":
            return len(v["features"])
    except (KeyError, TypeError):
        return 0
    return 0


def stats(samples):
    s = sorted(samples)
    n = len(s)
    idx = min(n - 1, n * 95 // 100)
    return s[0], s[n // 2], sum(s) // n, s[idx]


def row(run_id, ts, dataset, nbytes, lib, ver, warmup, trials, st, check_ok, status, ver_flags="-OO"):
    mn, med, mean, p95 = st
    mbps = (nbytes / med * 1e9 / 1048576.0) if med > 0 else 0.0
    pyver = env("PROFILE_LANG_VERSION", "")
    return (
        f"1,{run_id},{ts},json-dom,codec,{dataset},{nbytes},,{nbytes},python,{pyver},"
        f"{lib},{ver},{ver_flags},{warmup},{trials},{mn},{med},{mean},{p95},"
        f"{mbps:.1f},MB/s,{peak_rss_kb()},-1,{_ALLOC},-1,-1,{status},{str(check_ok).lower()},,"
    )


def skip_row(run_id, ts, dataset, nbytes, lib, reason):
    pyver = env("PROFILE_LANG_VERSION", "")
    return (
        f"1,{run_id},{ts},json-dom,codec,{dataset},{nbytes},,{nbytes},python,{pyver},"
        f"{lib},,,0,0,-1,-1,-1,-1,,,-1,-1,-1,-1,-1,skipped,,{reason},"
    )


def loader(lib):
    """Return (loads_callable, version) for a library, or None if unavailable."""
    try:
        if lib == "json":
            import json
            return (lambda b: json.loads(b), sys.version.split()[0])
        if lib == "orjson":
            orjson = importlib.import_module("orjson")
            return (orjson.loads, getattr(orjson, "__version__", ""))
        if lib == "msgspec":
            msgspec = importlib.import_module("msgspec")
            dec = msgspec.json.decode
            return (lambda b: dec(b), getattr(msgspec, "__version__", ""))
        if lib == "ujson":
            ujson = importlib.import_module("ujson")
            return (ujson.loads, getattr(ujson, "__version__", ""))
    except ImportError:
        return None
    return None


def main():
    global _ALLOC
    data_dir = os.environ.get("PROFILE_DATA_DIR")
    if not data_dir:
        sys.stderr.write("python/codec: PROFILE_DATA_DIR unset — skipping\n")
        return
    run_id = env("PROFILE_RUN_ID", "local")
    ts = env("PROFILE_RUN_TS", "")
    warmup = int(env("PROFILE_WARMUP", "3"))
    trials = int(env("PROFILE_TRIALS", "10"))

    libs = ["json", "orjson", "msgspec", "ujson"]
    out = []
    for dataset, fname, expect in FILES:
        path = os.path.join(data_dir, fname)
        try:
            with open(path, "rb") as f:
                raw = f.read()
        except OSError:
            continue  # dataset absent → skip
        n = len(raw)

        for lib in libs:
            ld = loader(lib)
            if ld is None:
                out.append(skip_row(run_id, ts, dataset, n, lib, f"{lib} not importable"))
                continue
            loads, ver = ld
            try:
                check = count(loads(raw), dataset) == expect
            except Exception:
                check = False
            _ALLOC = alloc_bytes(lambda: loads(raw))
            for _ in range(warmup):
                loads(raw)
            samples = []
            for _ in range(trials):
                t0 = time.perf_counter_ns()
                loads(raw)
                samples.append(time.perf_counter_ns() - t0)
            status = "ok" if check else "invalid"
            out.append(row(run_id, ts, dataset, n, lib, ver, warmup, trials,
                           stats(samples), check, status))

    sys.stdout.write("\n".join(out) + "\n")


if __name__ == "__main__":
    main()
